<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="LucianoRogerio" />

<meta name="date" content="2022-07-03" />

<title>PhenoData</title>

<script src="site_libs/header-attrs-2.21/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/pagedtable-1.1/css/pagedtable.css" rel="stylesheet" />
<script src="site_libs/pagedtable-1.1/js/pagedtable.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">NRCRI_2022GS</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">PhenoData</h1>
<h4 class="author">LucianoRogerio</h4>
<h4 class="date">2022-07-03</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2023-07-10
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 6
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 1
</p>
<p>
<strong>Knit directory:</strong> <code>NRCRI_2022GS/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguncommittedchanges">
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> <strong>R Markdown file:</strong> uncommitted
changes </a>
</p>
</div>
<div id="strongRMarkdownfilestronguncommittedchanges"
class="panel-collapse collapse">
<div class="panel-body">
<p>The R Markdown file has unstaged changes. To know which version of
the R Markdown file created these results, you’ll want to first commit
it to the Git repo. If you’re still working on the analysis, you can
ignore this warning. When you’re finished, you can run
<code>wflow_publish</code> to commit the R Markdown file and build the
HTML.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20230710code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20230710)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20230710code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20230710)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomLucianoRogerioNRCRI2022GStree0506c8a119bc858575c4c4b9448839c29aa2c22dtargetblank0506c8aa">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/LucianoRogerio/NRCRI_2022GS/tree/0506c8a119bc858575c4c4b9448839c29aa2c22d" target="_blank">0506c8a</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomLucianoRogerioNRCRI2022GStree0506c8a119bc858575c4c4b9448839c29aa2c22dtargetblank0506c8aa"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/LucianoRogerio/NRCRI_2022GS/tree/0506c8a119bc858575c4c4b9448839c29aa2c22d" target="_blank">0506c8a</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/
    Ignored:    data/2022-07-03T214144phenotype_download.csv

Untracked files:
    Untracked:  output/DosageMatrix_DCas22_7085RefPopImputed.rds
    Untracked:  output/NRCRI_CleanedTrialData_2022Jul09.rds
    Untracked:  output/NRCRI_ExptDesignsDetected_2022Jul09.rds
    Untracked:  output/OnlyChosen_germplasmName_to_FullSampleName_matches_NRCRI_2022Jul09.csv

Unstaged changes:
    Modified:   analysis/PhenoData.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/PhenoData.Rmd</code>) and HTML
(<code>docs/PhenoData.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/LucianoRogerio/NRCRI_2022GS/blob/0506c8a119bc858575c4c4b9448839c29aa2c22d/analysis/PhenoData.Rmd" target="_blank">0506c8a</a>
</td>
<td>
LucianoRogerio
</td>
<td>
2023-07-10
</td>
<td>
Fix Website
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="download-of-the-phenotypic-data-on-cassavabase"
class="section level2">
<h2>Download of the phenotypic data on Cassavabase</h2>
<p>Follow outlined GenomicPredictionChecklist and previous pipeline to
process cassavabase data for ultimate genomic prediction.</p>
<p>Below we will clean and format training data.</p>
<p>Inputs: “Raw” field trial data Expected outputs: “Cleaned” field
trial data</p>
<div id="cassavabase-download" class="section level3">
<h3>Cassavabase download</h3>
<p>Downloaded all NRCRI field trials, with all the traits using the <a
href="https://cassavabase.org/breeders/download">download tool</a>.</p>
<p>Yout need to create a list with all the NRCRI trials on the <a
href="https://cassavabase.org/breeders/search">wizard tool</a>
Cassavabase search wizard:</p>
<p>Selected all NRCRI trials currently available. Make a list. Named it
NRCRITrials_2022.</p>
<p>Go to Manage –&gt; <a
href="https://cassavabase.org/breeders/download">Download here</a>.
Download phenotypes (plot-basis only) and meta-data as CSV using the
corresponding boxes / drop-downs.</p>
<p>TRIED TO DOWNLOAD META-DATA, BUT DB IS GIVING “SERVER ERROR” This
happened last year too, must have something to do with NRCRI trials in
the list? Store flatfiles, unaltered in directory data/. Same error
noticed by Marnin.</p>
<p>Read DB data directly from the Cassavabase FTP server.</p>
<pre class="r"><code>library(genomicMateSelectR)</code></pre>
<pre><code>Loading required package: tibble</code></pre>
<pre><code>Loading required package: dplyr</code></pre>
<pre><code>
Attaching package: &#39;dplyr&#39;</code></pre>
<pre><code>The following objects are masked from &#39;package:stats&#39;:

    filter, lag</code></pre>
<pre><code>The following objects are masked from &#39;package:base&#39;:

    intersect, setdiff, setequal, union</code></pre>
<pre><code>Loading required package: tidyr</code></pre>
<pre><code>Loading required package: purrr</code></pre>
<pre class="r"><code>dbdata&lt;-genomicMateSelectR::readDBdata(phenotypeFile = here::here(&quot;data&quot;,
                                                                  &quot;2022-07-03T214144phenotype_download.csv&quot;))</code></pre>
</div>
<div id="group-and-select-trials-to-analyze" class="section level3">
<h3>Group and select trials to analyze</h3>
<p>Make TrialType Variable</p>
<pre class="r"><code>makeTrialTypeVarLuc &lt;- function(indata){
  # So far, this function is not very general
  # Can customize this or add lines to grab TrialTypes for each breeding program
  if(unique(indata$programName)==&quot;NRCRI&quot;){
    outdata&lt;-indata %&gt;%
      mutate(TrialType=ifelse(grepl(&quot;TP1&quot;,studyName,ignore.case = T),&quot;TP1&quot;,NA),
             TrialType=ifelse(grepl(&quot;TP2&quot;,studyName,ignore.case = T),&quot;TP2&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;C1a&quot;,studyName,ignore.case = T),&quot;C1a&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;C1b&quot;,studyName,ignore.case = T),&quot;C1b&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;C2a&quot;,studyName,ignore.case = T),&quot;C2a&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;C2b&quot;,studyName,ignore.case = T),&quot;C2b&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;C3a&quot;,studyName,ignore.case = T),&quot;C3a&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;C3b&quot;,studyName,ignore.case = T),&quot;C3b&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;clonal&quot;,studyName,ignore.case = T),&quot;CET&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;cet&quot;,studyName,ignore.case = T),&quot;CET&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;CET&quot;,studyName,ignore.case = T),&quot;CET&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;pyt&quot;,studyName,ignore.case = T),&quot;PYT&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;PYT&quot;,studyName,ignore.case = T),&quot;PYT&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;ayt&quot;,studyName,ignore.case = T),&quot;AYT&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;AYT&quot;,studyName,ignore.case = T),&quot;AYT&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;regional&quot;,studyName,ignore.case = T),&quot;UYT&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;uyt&quot;,studyName,ignore.case = T),&quot;UYT&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;UYT&quot;,studyName,ignore.case = T),&quot;UYT&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;NCRP&quot;,studyName) &amp; is.na(TrialType),&quot;NCRP&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;15nextgen60gs-cbUM|crossnblk|crossingblock&quot;,studyName,ignore.case = T) &amp;
                                !grepl(&quot;CET&quot;,studyName),
                              &quot;CrossingBlock&quot;,TrialType),
             TrialType=ifelse(grepl(&quot;seedling&quot;,studyName,ignore.case = T),NA,TrialType)) }

  return(outdata) }

dbdata&lt;-makeTrialTypeVarLuc(dbdata) 
dbdata %&gt;% 
  count(TrialType) %&gt;% rmarkdown::paged_table()</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["TrialType"],"name":[1],"type":["chr"],"align":["left"]},{"label":["n"],"name":[2],"type":["int"],"align":["right"]}],"data":[{"1":"AYT","2":"6018"},{"1":"C1b","2":"108"},{"1":"CET","2":"20566"},{"1":"CrossingBlock","2":"141"},{"1":"NCRP","2":"270"},{"1":"PYT","2":"13626"},{"1":"TP1","2":"5924"},{"1":"TP2","2":"6748"},{"1":"UYT","2":"3821"},{"1":"NA","2":"12789"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="remove-unclassified-trials" class="section level3">
<h3>Remove unclassified trials</h3>
<pre class="r"><code>dbdata %&lt;&gt;% 
    filter(!is.na(TrialType)) 
dbdata %&gt;% 
    group_by(programName) %&gt;% 
    summarize(N=n()) %&gt;% rmarkdown::paged_table()</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["programName"],"name":[1],"type":["chr"],"align":["left"]},{"label":["N"],"name":[2],"type":["int"],"align":["right"]}],"data":[{"1":"NRCRI","2":"57222"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="make-abbreviations-for-renaming" class="section level3">
<h3>Make abbreviations for renaming</h3>
<pre class="r"><code>traitabbrevs&lt;-tribble(~TraitAbbrev,~TraitName,
        &quot;CMD1S&quot;, &quot;cassava.mosaic.disease.severity.1.month.evaluation.CO_334.0000191&quot;,
        &quot;CMD3S&quot;, &quot;cassava.mosaic.disease.severity.3.month.evaluation.CO_334.0000192&quot;,
        &quot;CMD6S&quot;, &quot;cassava.mosaic.disease.severity.6.month.evaluation.CO_334.0000194&quot;,
        &quot;CMD9S&quot;, &quot;cassava.mosaic.disease.severity.9.month.evaluation.CO_334.0000193&quot;,
        &quot;CGM&quot;, &quot;Cassava.green.mite.severity.CO_334.0000033&quot;,
        &quot;CGMS1&quot;,&quot;cassava.green.mite.severity.first.evaluation.CO_334.0000189&quot;,
        &quot;CGMS2&quot;,&quot;cassava.green.mite.severity.second.evaluation.CO_334.0000190&quot;,
        &quot;DMC&quot;,&quot;dry.matter.content.percentage.CO_334.0000092&quot;,
        &quot;DMCsg&quot;,&quot;dry.matter.content.by.specific.gravity.method.CO_334.0000160&quot;,
        &quot;DRY&quot;, &quot;dry.yield.CO_334.0000014&quot;,
        &quot;IVig1-5&quot;, &quot;initial.plant.vigor.assessment.1.5.CO_334.0000220&quot;,
        &quot;IVig1-7&quot;, &quot;initial.vigor.assessment.1.7.CO_334.0000009&quot;,
        &quot;PLTHT&quot;,&quot;plant.height.measurement.in.cm.CO_334.0000018&quot;,
        &quot;BRNHT1&quot;,&quot;first.apical.branch.height.measurement.in.cm.CO_334.0000106&quot;,
        &quot;SHTWT&quot;,&quot;fresh.shoot.weight.measurement.in.kg.per.plot.CO_334.0000016&quot;,
        &quot;RTWT&quot;,&quot;fresh.storage.root.weight.per.plot.CO_334.0000012&quot;,
        &quot;RTNO&quot;,&quot;root.number.counting.CO_334.0000011&quot;,
        &quot;TCHART&quot;,&quot;total.carotenoid.by.chart.1.8.CO_334.0000161&quot;,
        &quot;NOHAV&quot;,&quot;plant.stands.harvested.counting.CO_334.0000010&quot;,
        &quot;StC&quot;, &quot;starch.content.percentage.CO_334.0000071&quot;)

traitabbrevs %&gt;% rmarkdown::paged_table()</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["TraitAbbrev"],"name":[1],"type":["chr"],"align":["left"]},{"label":["TraitName"],"name":[2],"type":["chr"],"align":["left"]}],"data":[{"1":"CMD1S","2":"cassava.mosaic.disease.severity.1.month.evaluation.CO_334.0000191"},{"1":"CMD3S","2":"cassava.mosaic.disease.severity.3.month.evaluation.CO_334.0000192"},{"1":"CMD6S","2":"cassava.mosaic.disease.severity.6.month.evaluation.CO_334.0000194"},{"1":"CMD9S","2":"cassava.mosaic.disease.severity.9.month.evaluation.CO_334.0000193"},{"1":"CGM","2":"Cassava.green.mite.severity.CO_334.0000033"},{"1":"CGMS1","2":"cassava.green.mite.severity.first.evaluation.CO_334.0000189"},{"1":"CGMS2","2":"cassava.green.mite.severity.second.evaluation.CO_334.0000190"},{"1":"DMC","2":"dry.matter.content.percentage.CO_334.0000092"},{"1":"DMCsg","2":"dry.matter.content.by.specific.gravity.method.CO_334.0000160"},{"1":"DRY","2":"dry.yield.CO_334.0000014"},{"1":"IVig1-5","2":"initial.plant.vigor.assessment.1.5.CO_334.0000220"},{"1":"IVig1-7","2":"initial.vigor.assessment.1.7.CO_334.0000009"},{"1":"PLTHT","2":"plant.height.measurement.in.cm.CO_334.0000018"},{"1":"BRNHT1","2":"first.apical.branch.height.measurement.in.cm.CO_334.0000106"},{"1":"SHTWT","2":"fresh.shoot.weight.measurement.in.kg.per.plot.CO_334.0000016"},{"1":"RTWT","2":"fresh.storage.root.weight.per.plot.CO_334.0000012"},{"1":"RTNO","2":"root.number.counting.CO_334.0000011"},{"1":"TCHART","2":"total.carotenoid.by.chart.1.8.CO_334.0000161"},{"1":"NOHAV","2":"plant.stands.harvested.counting.CO_334.0000010"},{"1":"StC","2":"starch.content.percentage.CO_334.0000071"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Run function renameAndSelectCols() to rename columns and remove
everything unecessary</p>
<pre class="r"><code>dbdata&lt;-renameAndSelectCols(traitabbrevs,indata=dbdata,customColsToKeep = c(&quot;TrialType&quot;,&quot;observationUnitName&quot;))</code></pre>
<pre><code>Joining with `by = join_by(TraitName)`</code></pre>
</div>
<div id="removing-unexpected-trait-values---quality-control"
class="section level3">
<h3>Removing unexpected trait values - Quality Control</h3>
<pre class="r"><code>dbdata&lt;-dbdata %&gt;% 
  mutate(CMD1S=ifelse(CMD1S&lt;1 | CMD1S&gt;5,NA,CMD1S),
         CMD3S=ifelse(CMD3S&lt;1 | CMD3S&gt;5,NA,CMD3S),
         CMD6S=ifelse(CMD6S&lt;1 | CMD6S&gt;5,NA,CMD6S),
         CMD9S=ifelse(CMD9S&lt;1 | CMD9S&gt;5,NA,CMD9S),
         CGM=ifelse(CGM&lt;1 | CGM&gt;5,NA,CGM),
         CGMS1=ifelse(CGMS1&lt;1 | CGMS1&gt;5,NA,CGMS1),
         CGMS2=ifelse(CGMS2&lt;1 | CGMS2&gt;5,NA,CGMS2),
         DMC=ifelse(DMC&gt;100 | DMC&lt;=0,NA,DMC),
         DMCsg=ifelse(DMCsg&gt;100 | DMCsg&lt;=0,NA,DMCsg),
         DRY = ifelse(DRY==0 | NOHAV==0 | is.na(NOHAV),NA,DRY),
         `IVig1-5`=ifelse(`IVig1-5`&lt;1 | `IVig1-5`&gt;5,NA,`IVig1-5`),
         `IVig1-7`=ifelse(`IVig1-7`&lt;1 | `IVig1-7`&gt;7,NA,`IVig1-7`),
         RTWT=ifelse(RTWT==0 | NOHAV==0 | is.na(NOHAV),NA,RTWT),
         SHTWT=ifelse(SHTWT==0 | NOHAV==0 | is.na(NOHAV),NA,SHTWT),
         NOHAV=ifelse(NOHAV==0,NA,NOHAV),
         NOHAV=ifelse(NOHAV&gt;42,NA,NOHAV),
         StC=ifelse(StC&gt;100 | StC&lt;=0,NA,StC))</code></pre>
</div>
</div>
<div id="post-qc-traits" class="section level2">
<h2>Post-QC traits</h2>
<div id="harvest-index" class="section level3">
<h3>Harvest index</h3>
<pre class="r"><code>dbdata&lt;-dbdata %&gt;% 
    mutate(HI=RTWT/(RTWT+SHTWT))</code></pre>
</div>
<div id="unit-area-traits" class="section level3">
<h3>Unit area traits</h3>
<p>I anticipate this will not be necessary as it will be computed before
or during data upload.</p>
<p>For calculating fresh root yield:</p>
<ol style="list-style-type: decimal">
<li>PlotSpacing: Area in m2 per plant. plotWidth and plotLength metadata
would hypothetically provide this info, but is missing for vast majority
of trials. Therefore, use info from Fola.</li>
<li>maxNOHAV: Instead of ExpectedNOHAV. Need to know the max number of
plants in the area harvested. For some trials, only the inner (or “net”)
plot is harvested, therefore the PlantsPerPlot meta-variable will not
suffice. Besides, the PlantsPerPlot information is missing for the vast
majority of trials. Instead, use observed max(NOHAV) for each trial. We
use this plus the PlotSpacing to calc. the area over which the RTWT was
measured. During analysis, variation in the actual number of plants
harvested will be accounted for.</li>
</ol>
<pre class="r"><code>dbdata&lt;-dbdata %&gt;% 
    mutate(PlotSpacing=ifelse(programName!=&quot;IITA&quot;,1,
                              ifelse(studyYear&lt;2013,1,
                              ifelse(TrialType %in% c(&quot;CET&quot;,&quot;GeneticGain&quot;,&quot;ExpCET&quot;),1,0.8))))
maxNOHAV_byStudy&lt;-dbdata %&gt;% 
  group_by(programName,locationName,studyYear,studyName,studyDesign, TrialType) %&gt;% 
  dplyr::summarise(MaxNOHAV=max(NOHAV, na.rm=T)) %&gt;% 
  ungroup() %&gt;% 
  mutate(MaxNOHAV=ifelse(MaxNOHAV==&quot;-Inf&quot;,NA,MaxNOHAV),
         MaxNoHAV=ifelse(TrialType==&quot;CET&quot;,
                         yes = 5,
                         no = ifelse(TrialType==&quot;PYT&quot;,
                                     yes = 10,
                                     no = ifelse(TrialType==&quot;AYT&quot;,
                                                 yes = 20,
                                                 no = ifelse(TrialType==&quot;UYT&quot;,
                                                             yes = 36,
                                                             no = MaxNOHAV)))))</code></pre>
<pre><code>Warning: There were 171 warnings in `dplyr::summarise()`.
The first warning was:
ℹ In argument: `MaxNOHAV = max(NOHAV, na.rm = T)`.
ℹ In group 2: `programName = &quot;NRCRI&quot;`, `locationName = &quot;Ago-Owu&quot;`, `studyYear =
  2020`, `studyName = &quot;20AYTintroLatAmeAgo-Owu&quot;`, `studyDesign = &quot;Alpha&quot;`,
  `TrialType = &quot;AYT&quot;`.
Caused by warning in `max()`:
! no non-missing arguments to max; returning -Inf
ℹ Run `dplyr::last_dplyr_warnings()` to see the 170 remaining warnings.</code></pre>
<pre><code>`summarise()` has grouped output by &#39;programName&#39;, &#39;locationName&#39;, &#39;studyYear&#39;,
&#39;studyName&#39;, &#39;studyDesign&#39;. You can override using the `.groups` argument.</code></pre>
<pre class="r"><code>write.csv(maxNOHAV_byStudy %&gt;% arrange(studyYear),file=here::here(&quot;output&quot;,&quot;maxNOHAV_byStudy.csv&quot;), row.names = F)</code></pre>
<p>This part is copied from Marnin 2021 NRCRI prediction</p>
<p>Previously, I took these values as is. I am unsatisfied with that.
The trial number is small enough I’m going to curate manually below. I
hope this gives better yield results.</p>
<pre class="r"><code>maxNOHAV_byStudy %&lt;&gt;%
  mutate(MaxNOHAV=ifelse(studyName==&quot;18C2acrossingblockCETubiaja&quot;,8,MaxNOHAV),
         MaxNOHAV=ifelse(studyName==&quot;13TP1CET518kano&quot;,5,MaxNOHAV),
         MaxNOHAV=ifelse(studyName==&quot;17C1aAYTGSkano&quot;,10,MaxNOHAV),
         MaxNOHAV=ifelse(studyName==&quot;18C1bAYTGSOtobi&quot;,10,MaxNOHAV),
         MaxNOHAV=ifelse(studyName==&quot;16C1aCETnonGSOtobi&quot;,5,MaxNOHAV),
         MaxNOHAV=ifelse(studyName==&quot;17C1bCETkano&quot;,5,MaxNOHAV),
         MaxNOHAV=ifelse(studyName==&quot;16C1aCETnonGSOtobi&quot;,5,MaxNOHAV),
         MaxNOHAV=ifelse(studyName==&quot;18C1bAYTGSset2umu&quot;,10,MaxNOHAV))</code></pre>
<p>Now back to the standart workflow</p>
<pre class="r"><code>dbdata&lt;-left_join(dbdata,maxNOHAV_byStudy) %&gt;% 
  mutate(RTWT=ifelse(NOHAV&gt;MaxNOHAV,NA,RTWT),
         SHTWT=ifelse(NOHAV&gt;MaxNOHAV,NA,SHTWT),
         HI=ifelse(NOHAV&gt;MaxNOHAV,NA,HI),
         FYLD=RTWT/(MaxNOHAV*PlotSpacing)*10,
         DRYLD=FYLD*(DMC/100),
         DRYsgLD=FYLD*(DMCsg/100),
         logFYLD=log(FYLD),
         logDRYLD=log(DRYLD),
         logDRYsgLD=log(DRYsgLD),
         logTOPYLD=log(SHTWT/(MaxNOHAV*PlotSpacing)*10),
         PropNOHAV=NOHAV/MaxNOHAV) </code></pre>
<pre><code>Joining with `by = join_by(studyYear, programName, locationName, studyName,
studyDesign, TrialType)`</code></pre>
<pre class="r"><code># remove non transformed / per-plot (instead of per area) traits
dbdata %&lt;&gt;% select(-RTWT,-SHTWT,-FYLD,-DRYLD,-DRYsgLD)</code></pre>
</div>
<div id="season-wide-mean-disease-severity" class="section level3">
<h3>Season-wide mean disease severity</h3>
<pre class="r"><code>dbdata&lt;-dbdata %&gt;% 
  mutate(MCMDS=rowMeans(.[,colnames(.) %in% c(&quot;CMD1S&quot;,&quot;CMD3S&quot;,&quot;CMD6S&quot;,&quot;CMD9S&quot;)], na.rm = T)) %&gt;% 
  select(-any_of(c(&quot;CMD1S&quot;,&quot;CMD3S&quot;,&quot;CMD6S&quot;,&quot;CMD9S&quot;)))

write.table(unique(dbdata$germplasmName), file = here::here(&quot;data&quot;, &quot;NRCRIClones.txt&quot;),
            quote = F, row.names = F, col.names = F)</code></pre>
</div>
</div>
<div id="user-input-assign-genos-to-phenos" class="section level2">
<h2>[User input] Assign genos to phenos</h2>
<p>This step is mostly copy-pasted from previous processing of IITA- and
NRCRI-specific data.</p>
<p>Uses 4 flat files, which are available e.g. here. Specifically,
<code>IITA_GBStoPhenoMaster_33018.csv</code>,
<code>GBSdataMasterList_31818.csv</code> and
<code>NRCRI_GBStoPhenoMaster_40318.csv</code> and
<code>chr1_RefPanelAndGSprogeny_ReadyForGP_72719.fam</code>. I copy them
to the <code>data/</code> sub-directory for the current analysis.</p>
<p>I downloaded all the Genotyping plates related to the DArT report
<code>DCas-22_7085</code> (30 genotyping plates).</p>
<p>In addition, DArT-only samples are now expected to also have
phenotypes. Therefore, checking for matches in new flatfiles, deposited
in the data/ (see code below).</p>
<pre class="r"><code>library(tidyverse); library(magrittr)
library(here)

DArTNames2018 &lt;- dir(here::here(&quot;data&quot;, &quot;GenotypingPlates&quot;, &quot;C2&quot;))
DArTNames2020 &lt;- dir(here::here(&quot;data&quot;, &quot;GenotypingPlates&quot;, &quot;C3&quot;))
DArTNames2022 &lt;- dir(here::here(&quot;data&quot;, &quot;GenotypingPlates&quot;, &quot;C4&quot;))

GPDArT2018 &lt;- tibble()
for(i in DArTNames2018){
file &lt;- read.table(file = here::here(&quot;data&quot;, &quot;GenotypingPlates&quot;, &quot;C2&quot;, i), sep = &quot;,&quot;,
                   header = T)

GPDArT2018 &lt;- rbind(GPDArT2018, file)
rm(file)
}

GPDArT2020 &lt;- tibble()
for(i in DArTNames2020){
file &lt;- read.table(file = here::here(&quot;data&quot;, &quot;GenotypingPlates&quot;, &quot;C3&quot;, i), sep = &quot;,&quot;,
                   header = T)

GPDArT2020 &lt;- rbind(GPDArT2020, file)
rm(file)
}

GPDArT2022 &lt;- tibble()
for(i in DArTNames2022){
file &lt;- read.table(file = here::here(&quot;data&quot;, &quot;GenotypingPlates&quot;, &quot;C4&quot;, i), sep = &quot;,&quot;,
                   header = T)

GPDArT2022 &lt;- rbind(GPDArT2022, file)
rm(file)
}


GPDArT &lt;- rbind(GPDArT2018, GPDArT2020, GPDArT2022)
write.table(x = GPDArT, file = here::here(&quot;data&quot;, &quot;NRCRIDArT2022Clones.csv&quot;),
            sep = &quot;,&quot;, quote = F, row.names = F)

saveRDS(dbdata, file = here::here(&quot;data&quot;, &quot;dbdata.rds&quot;))</code></pre>
<pre class="r"><code>dbdata &lt;- readRDS(file = here::here(&quot;data&quot;, &quot;dbdata.rds&quot;))

gbs2phenoMaster&lt;-dbdata %&gt;% 
  select(germplasmName) %&gt;% 
  distinct %&gt;% 
  left_join(read.csv(here::here(&quot;data&quot;,&quot;NRCRI_GBStoPhenoMaster_40318.csv&quot;), 
                     stringsAsFactors = F)) %&gt;% 
  mutate(FullSampleName=ifelse(grepl(&quot;C2a&quot;,germplasmName,ignore.case = T) &amp; 
                                 is.na(FullSampleName),germplasmName,FullSampleName)) %&gt;% 
  filter(!is.na(FullSampleName)) %&gt;% 
  select(germplasmName,FullSampleName) %&gt;% 
  bind_rows(dbdata %&gt;% 
              select(germplasmName) %&gt;% 
              distinct %&gt;% 
              left_join(read.csv(here::here(&quot;data&quot;,&quot;IITA_GBStoPhenoMaster_33018.csv&quot;), 
                                 stringsAsFactors = F)) %&gt;% 
              filter(!is.na(FullSampleName)) %&gt;% 
              select(germplasmName,FullSampleName)) %&gt;% 
  bind_rows(dbdata %&gt;% 
              select(germplasmName) %&gt;% 
              distinct %&gt;% 
              left_join(read.csv(here::here(&quot;data&quot;,&quot;GBSdataMasterList_31818.csv&quot;), 
                                 stringsAsFactors = F) %&gt;% 
                          select(DNASample,FullSampleName) %&gt;% 
                          rename(germplasmName=DNASample)) %&gt;% 
              filter(!is.na(FullSampleName)) %&gt;% 
              select(germplasmName,FullSampleName)) %&gt;% 
  bind_rows(dbdata %&gt;% 
              select(germplasmName) %&gt;% 
              distinct %&gt;% 
              mutate(germplasmSynonyms=ifelse(grepl(&quot;^UG&quot;,germplasmName,ignore.case = T),
                                              gsub(&quot;UG&quot;,&quot;Ug&quot;,germplasmName),germplasmName)) %&gt;% 
              left_join(read.csv(here::here(&quot;data&quot;,&quot;GBSdataMasterList_31818.csv&quot;), 
                                 stringsAsFactors = F) %&gt;% 
                          select(DNASample,FullSampleName) %&gt;% 
                          rename(germplasmSynonyms=DNASample)) %&gt;% 
              filter(!is.na(FullSampleName)) %&gt;% 
              select(germplasmName,FullSampleName)) %&gt;%
    bind_rows(dbdata %&gt;% 
              select(germplasmName) %&gt;% 
              distinct %&gt;% 
              left_join(read.csv(here::here(&quot;data&quot;,&quot;NRCRIDArT2022Clones.csv&quot;), 
                                 stringsAsFactors = F) %&gt;% 
                          select(exported_tissue_sample_name, accession_name) %&gt;% 
                          rename(FullSampleName = exported_tissue_sample_name,
                                 germplasmName = accession_name)) %&gt;% 
              filter(!is.na(FullSampleName)) %&gt;% 
              select(germplasmName,FullSampleName)) %&gt;% 
  bind_rows(dbdata %&gt;% 
              select(germplasmName) %&gt;% 
              distinct %&gt;% 
              mutate(germplasmSynonyms=ifelse(grepl(&quot;^TZ&quot;,germplasmName,
                                                    ignore.case = T),
                                              gsub(&quot;TZ&quot;,&quot;&quot;,germplasmName),germplasmName)) %&gt;% 
              left_join(read.csv(here::here(&quot;data&quot;,&quot;GBSdataMasterList_31818.csv&quot;), 
                                 stringsAsFactors = F) %&gt;% 
                          select(DNASample,FullSampleName) %&gt;% 
                          rename(germplasmSynonyms=DNASample)) %&gt;% 
              filter(!is.na(FullSampleName)) %&gt;%
              select(germplasmName,FullSampleName)) %&gt;% 
  distinct %&gt;% 
  left_join(read.csv(here::here(&quot;data&quot;,&quot;GBSdataMasterList_31818.csv&quot;), 
                     stringsAsFactors = F) %&gt;% 
              select(FullSampleName,OrigKeyFile,Institute) %&gt;% 
              rename(OriginOfSample=Institute)) %&gt;% 
  mutate(OrigKeyFile=ifelse(grepl(&quot;C2a&quot;,germplasmName,ignore.case = T),
                            ifelse(is.na(OrigKeyFile),&quot;LavalGBS&quot;,OrigKeyFile),
                            OrigKeyFile),
         OriginOfSample=ifelse(grepl(&quot;C2a&quot;,germplasmName,ignore.case = T),
                               ifelse(is.na(OriginOfSample),&quot;NRCRI&quot;,OriginOfSample),
                               OriginOfSample))</code></pre>
<pre><code>Joining with `by = join_by(germplasmName)`
Joining with `by = join_by(germplasmName)`
Joining with `by = join_by(germplasmName)`
Joining with `by = join_by(germplasmSynonyms)`
Joining with `by = join_by(germplasmName)`
Joining with `by = join_by(germplasmSynonyms)`
Joining with `by = join_by(FullSampleName)`</code></pre>
<pre class="r"><code>## NEW: check for germName-DArT name matches
germNamesWithoutGBSgenos&lt;-dbdata %&gt;% 
  select(programName,germplasmName) %&gt;% 
  distinct %&gt;% 
  left_join(gbs2phenoMaster) %&gt;% 
  filter(is.na(FullSampleName)) %&gt;% 
  select(-FullSampleName)</code></pre>
<pre><code>Joining with `by = join_by(germplasmName)`</code></pre>
<pre class="r"><code>## NEW: check for germName-DArT name matches
germNamesWithoutGBSgenos&lt;-dbdata %&gt;% 
  select(programName,germplasmName) %&gt;% 
  distinct %&gt;% 
  left_join(gbs2phenoMaster) %&gt;% 
  filter(is.na(FullSampleName)) %&gt;% 
  select(-FullSampleName)</code></pre>
<pre><code>Joining with `by = join_by(germplasmName)`</code></pre>
<pre class="r"><code>germNamesWithDArT&lt;-germNamesWithoutGBSgenos %&gt;% 
  inner_join(read.table(here::here(&quot;data&quot;,&quot;chr1_RefPanelAndGSprogeny_ReadyForGP_72719.fam&quot;), 
                        header = F, stringsAsFactors = F)$V2 %&gt;% 
               grep(&quot;TMS16|TMS17|TMS18|TMS19|TMS20&quot;,.,value = T, ignore.case = T) %&gt;% 
               tibble(dartName=.) %&gt;% 
               separate(dartName,c(&quot;germplasmName&quot;,&quot;dartID&quot;),&quot;_&quot;,extra = &#39;merge&#39;,remove = F)) %&gt;% 
  group_by(germplasmName) %&gt;% 
  slice(1) %&gt;% 
  ungroup() %&gt;% 
  rename(FullSampleName=dartName) %&gt;% 
  mutate(OrigKeyFile=&quot;DArTseqLD&quot;, OriginOfSample=&quot;IITA&quot;) %&gt;% 
  select(-dartID)</code></pre>
<pre><code>Joining with `by = join_by(germplasmName)`</code></pre>
<pre class="r"><code>print(paste0(nrow(germNamesWithDArT),&quot; germNames with DArT-only genos&quot;))</code></pre>
<pre><code>[1] &quot;15 germNames with DArT-only genos&quot;</code></pre>
<pre class="r"><code># first, filter to just program-DNAorigin matches
germNamesWithGenos&lt;-dbdata %&gt;% 
  select(programName,germplasmName) %&gt;% 
  distinct %&gt;% 
  left_join(gbs2phenoMaster) %&gt;% 
  filter(!is.na(FullSampleName))</code></pre>
<pre><code>Joining with `by = join_by(germplasmName)`</code></pre>
<pre class="r"><code>print(paste0(nrow(germNamesWithGenos),&quot; germNames with GBS genos&quot;))</code></pre>
<pre><code>[1] &quot;5225 germNames with GBS genos&quot;</code></pre>
<pre class="r"><code># program-germNames with locally sourced GBS samples
germNamesWithGenos_HasLocalSourcedGBS&lt;-germNamesWithGenos %&gt;% 
  filter(programName==OriginOfSample) %&gt;% 
  select(programName,germplasmName) %&gt;% 
  semi_join(germNamesWithGenos,.) %&gt;% 
  group_by(programName,germplasmName) %&gt;% # select one DNA per germplasmName per program
  slice(1) %&gt;% ungroup() </code></pre>
<pre><code>Joining with `by = join_by(programName, germplasmName)`</code></pre>
<pre class="r"><code>print(paste0(nrow(germNamesWithGenos_HasLocalSourcedGBS),&quot; germNames with local GBS genos&quot;))</code></pre>
<pre><code>[1] &quot;2989 germNames with local GBS genos&quot;</code></pre>
<pre class="r"><code># the rest (program-germNames) with GBS but coming from a different breeding program
germNamesWithGenos_NoLocalSourcedGBS&lt;-germNamesWithGenos %&gt;% 
  filter(programName==OriginOfSample) %&gt;% 
  select(programName,germplasmName) %&gt;% 
  anti_join(germNamesWithGenos,.) %&gt;% 
  # select one DNA per germplasmName per program
  group_by(programName,germplasmName) %&gt;% 
  slice(1) %&gt;% ungroup() </code></pre>
<pre><code>Joining with `by = join_by(programName, germplasmName)`</code></pre>
<pre class="r"><code>print(paste0(nrow(germNamesWithGenos_NoLocalSourcedGBS),&quot; germNames without local GBS genos&quot;))</code></pre>
<pre><code>[1] &quot;978 germNames without local GBS genos&quot;</code></pre>
<pre class="r"><code>genosForPhenos&lt;-bind_rows(germNamesWithGenos_HasLocalSourcedGBS,
                        germNamesWithGenos_NoLocalSourcedGBS) %&gt;% 
  bind_rows(germNamesWithDArT)

print(paste0(nrow(genosForPhenos),&quot; total germNames with genos either GBS or DArT&quot;))</code></pre>
<pre><code>[1] &quot;3982 total germNames with genos either GBS or DArT&quot;</code></pre>
<pre class="r"><code>dbdata %&lt;&gt;% 
    left_join(genosForPhenos) </code></pre>
<pre><code>Joining with `by = join_by(programName, germplasmName)`</code></pre>
<pre class="r"><code># Create a new identifier, GID
## Equals the value SNP data name (FullSampleName) 
## else germplasmName if no SNP data
dbdata %&lt;&gt;% 
  mutate(GID=ifelse(is.na(FullSampleName),germplasmName,FullSampleName))

#  going to check against SNP data
snps&lt;-readRDS(file = here::here(&quot;output&quot;,
                                &quot;DosageMatrix_DCas22_7085RefPopImputed.rds&quot;))
rownames_snps&lt;-rownames(snps); rm(snps); gc()</code></pre>
<pre><code>          used (Mb) gc trigger   (Mb) limit (Mb)  max used   (Mb)
Ncells 1068205 57.1    1803035   96.3         NA   1803035   96.3
Vcells 6964727 53.2  767738477 5857.4      16384 798370683 6091.1</code></pre>
<pre class="r"><code>dbdata %&gt;% 
  distinct(GID,germplasmName,FullSampleName) %&gt;% 
  semi_join(tibble(GID=rownames_snps)) %&gt;% nrow() #</code></pre>
<pre><code>Joining with `by = join_by(GID)`</code></pre>
<pre><code>[1] 1397</code></pre>
<pre class="r"><code>dbdata %&gt;% 
  distinct(GID,germplasmName,FullSampleName) %&gt;% 
  semi_join(tibble(GID=rownames_snps)) %&gt;% 
  filter(grepl(&quot;C1&quot;,GID,ignore.case = F)) # 55 C1 clones currently match</code></pre>
<pre><code>Joining with `by = join_by(GID)`</code></pre>
<pre><code># A tibble: 55 × 3
   GID                            germplasmName FullSampleName                
   &lt;chr&gt;                          &lt;chr&gt;         &lt;chr&gt;                         
 1 F116C1bP006:CA7RRANXX:3:499841 F116C1bP006   F116C1bP006:CA7RRANXX:3:499841
 2 F11C1bP001:CABV7ANXX:5:503891  F11C1bP001    F11C1bP001:CABV7ANXX:5:503891 
 3 F123C1bP002:CA7RRANXX:4:499895 F123C1bP002   F123C1bP002:CA7RRANXX:4:499895
 4 F123C1bP006:CA7RRANXX:4:499899 F123C1bP006   F123C1bP006:CA7RRANXX:4:499899
 5 F12C1bP018:CABV7ANXX:5:503932  F12C1bP018    F12C1bP018:CABV7ANXX:5:503932 
 6 F135C1bP001:CABJYANXX:5:505238 F135C1bP001   F135C1bP001:CABJYANXX:5:505238
 7 F136C1bP002:CABJYANXX:5:505240 F136C1bP002   F136C1bP002:CABJYANXX:5:505240
 8 F140C1bP001:CABJYANXX:5:505270 F140C1bP001   F140C1bP001:CABJYANXX:5:505270
 9 F152C1bP010:CA7RRANXX:5:499961 F152C1bP010   F152C1bP010:CA7RRANXX:5:499961
10 F152C1bP014:CA7RRANXX:5:499965 F152C1bP014   F152C1bP014:CA7RRANXX:5:499965
# ℹ 45 more rows</code></pre>
<pre class="r"><code>dbdata %&gt;% 
  distinct(GID,germplasmName,FullSampleName) %&gt;% 
  semi_join(tibble(GID=rownames_snps)) %&gt;% 
  filter(grepl(&quot;C2&quot;,GID,ignore.case = F)) # 57 C2 clones either</code></pre>
<pre><code>Joining with `by = join_by(GID)`</code></pre>
<pre><code># A tibble: 57 × 3
   GID            germplasmName  FullSampleName
   &lt;chr&gt;          &lt;chr&gt;          &lt;chr&gt;         
 1 NR17F15C2aP011 NR17F15C2aP011 NR17F15C2aP011
 2 NR17F1C2aP002  NR17F1C2aP002  NR17F1C2aP002 
 3 NR17F1C2aP003  NR17F1C2aP003  NR17F1C2aP003 
 4 NR17F1C2aP005  NR17F1C2aP005  NR17F1C2aP005 
 5 NR17F1C2aP006  NR17F1C2aP006  NR17F1C2aP006 
 6 NR17F1C2aP007  NR17F1C2aP007  NR17F1C2aP007 
 7 NR17F1C2aP008  NR17F1C2aP008  NR17F1C2aP008 
 8 NR17F1C2aP009  NR17F1C2aP009  NR17F1C2aP009 
 9 NR17F1C2aP010  NR17F1C2aP010  NR17F1C2aP010 
10 NR17F1C2aP011  NR17F1C2aP011  NR17F1C2aP011 
# ℹ 47 more rows</code></pre>
<pre class="r"><code>dbdata %&gt;% 
  distinct(GID,germplasmName,FullSampleName) %&gt;% 
  semi_join(tibble(GID=rownames_snps)) %&gt;% 
  filter(grepl(&quot;C3&quot;,GID,ignore.case = F)) # no C3 clones either</code></pre>
<pre><code>Joining with `by = join_by(GID)`</code></pre>
<pre><code># A tibble: 0 × 3
# ℹ 3 variables: GID &lt;chr&gt;, germplasmName &lt;chr&gt;, FullSampleName &lt;chr&gt;</code></pre>
<pre class="r"><code>dbdata %&gt;% 
  distinct(GID,germplasmName,FullSampleName) %&gt;% 
  anti_join(tibble(GID=rownames_snps)) %&gt;% 
  filter(grepl(&quot;C1|C2|C3&quot;,GID,ignore.case = T)) # definitely there are both C1, C2, and C3 phenotypes</code></pre>
<pre><code>Joining with `by = join_by(GID)`</code></pre>
<pre><code># A tibble: 3,226 × 3
   GID             germplasmName   FullSampleName
   &lt;chr&gt;           &lt;chr&gt;           &lt;chr&gt;         
 1 NR16F100C1bP001 NR16F100C1bP001 &lt;NA&gt;          
 2 NR16F100C1bP002 NR16F100C1bP002 &lt;NA&gt;          
 3 NR16F104C1bP001 NR16F104C1bP001 &lt;NA&gt;          
 4 NR16F105C1bP001 NR16F105C1bP001 &lt;NA&gt;          
 5 NR16F105C1bP002 NR16F105C1bP002 &lt;NA&gt;          
 6 NR16F106C1bP001 NR16F106C1bP001 &lt;NA&gt;          
 7 NR16F107C1bP002 NR16F107C1bP002 &lt;NA&gt;          
 8 NR16F107C1bP004 NR16F107C1bP004 &lt;NA&gt;          
 9 NR16F108C1bP001 NR16F108C1bP001 &lt;NA&gt;          
10 NR16F109C1bP001 NR16F109C1bP001 &lt;NA&gt;          
# ℹ 3,216 more rows</code></pre>
<pre class="r"><code># and there are C1, C2, and C3 genotypes
rownames_snps %&gt;% grep(&quot;C1&quot;,.,value = T,ignore.case = T) %&gt;% length # [1] </code></pre>
<pre><code>[1] 1726</code></pre>
<pre class="r"><code>rownames_snps %&gt;% grep(&quot;C2&quot;,.,value = T,ignore.case = T) %&gt;% length # [1]  </code></pre>
<pre><code>[1] 57</code></pre>
<pre class="r"><code>rownames_snps %&gt;% grep(&quot;C3&quot;,.,value = T,ignore.case = T) %&gt;% length # [1]  </code></pre>
<pre><code>[1] 829</code></pre>
<pre class="r"><code>germ2snps&lt;-dbdata %&gt;% 
  distinct(germplasmName,FullSampleName) %&gt;% 
  semi_join(tibble(FullSampleName=rownames_snps)) %&gt;% 
  bind_rows(dbdata %&gt;% 
              distinct(germplasmName,FullSampleName) %&gt;% 
              anti_join(tibble(FullSampleName=rownames_snps)) %&gt;% 
              filter(grepl(&quot;c1a&quot;,germplasmName,ignore.case = T)) %&gt;% 
              select(-FullSampleName) %&gt;% 
              left_join(tibble(FullSampleName=rownames_snps) %&gt;% 
                          filter(grepl(&quot;c1a&quot;,FullSampleName,ignore.case = T)) %&gt;% 
                          separate(FullSampleName,c(&quot;dartID&quot;,&quot;germplasmName&quot;),&quot;\\.\\.\\.&quot;,extra = &#39;merge&#39;,remove = F) %&gt;% 
                          select(-dartID))) %&gt;% 
  bind_rows(dbdata %&gt;% 
              distinct(germplasmName,FullSampleName) %&gt;% 
              anti_join(tibble(FullSampleName=rownames_snps)) %&gt;% 
              filter(grepl(&quot;C1b&quot;,germplasmName,ignore.case = T)) %&gt;% 
              filter(grepl(&quot;NR16C1b&quot;,germplasmName,ignore.case = T)) %&gt;% 
              select(-FullSampleName) %&gt;%
              left_join(tibble(FullSampleName=rownames_snps) %&gt;% 
                          filter(grepl(&quot;c1b&quot;,FullSampleName,ignore.case = T)) %&gt;% 
                          separate(FullSampleName,c(&quot;germplasmName&quot;,&quot;GBS_ID&quot;),&quot;:&quot;,extra = &#39;merge&#39;,remove = F) %&gt;% 
                          select(-GBS_ID) %&gt;% 
                          mutate(germplasmName=gsub(&quot;C1b&quot;,&quot;&quot;,germplasmName),
                                 germplasmName=paste0(&quot;NR16C1b&quot;,germplasmName)))) %&gt;% 
  bind_rows(dbdata %&gt;% 
              distinct(germplasmName,FullSampleName) %&gt;% 
              anti_join(tibble(FullSampleName=rownames_snps)) %&gt;% 
              filter(grepl(&quot;C1b&quot;,germplasmName,ignore.case = T)) %&gt;% 
              filter(!grepl(&quot;NR16C1b&quot;,germplasmName,ignore.case = T)) %&gt;% 
              select(-FullSampleName) %&gt;% 
              left_join(tibble(FullSampleName=rownames_snps) %&gt;% 
                          filter(grepl(&quot;c1b&quot;,FullSampleName,ignore.case = T)) %&gt;% 
                          separate(FullSampleName,c(&quot;germplasmName&quot;,&quot;GBS_ID&quot;),&quot;:&quot;,extra = &#39;merge&#39;,remove = F) %&gt;% 
                          select(-GBS_ID) %&gt;% 
                          mutate(germplasmName=paste0(&quot;NR16&quot;,germplasmName)))) %&gt;% 
  bind_rows(dbdata %&gt;% 
              distinct(germplasmName,FullSampleName) %&gt;% 
              anti_join(tibble(FullSampleName=rownames_snps)) %&gt;% 
              filter(grepl(&quot;c2&quot;,germplasmName,ignore.case = T)) %&gt;% 
              select(-FullSampleName) %&gt;% 
              left_join(tibble(FullSampleName=rownames_snps) %&gt;% 
                          filter(grepl(&quot;c2&quot;,FullSampleName,ignore.case = T),
                                 grepl(&quot;\\.\\.\\.&quot;,FullSampleName)) %&gt;% 
                          separate(FullSampleName,c(&quot;dartID&quot;,&quot;germplasmName&quot;),&quot;\\.\\.\\.&quot;,extra = &#39;merge&#39;,remove = F) %&gt;% 
                          select(-dartID))) %&gt;% 
  distinct</code></pre>
<pre><code>Joining with `by = join_by(FullSampleName)`
Joining with `by = join_by(FullSampleName)`
Joining with `by = join_by(germplasmName)`
Joining with `by = join_by(FullSampleName)`
Joining with `by = join_by(germplasmName)`
Joining with `by = join_by(FullSampleName)`
Joining with `by = join_by(germplasmName)`
Joining with `by = join_by(FullSampleName)`
Joining with `by = join_by(germplasmName)`</code></pre>
<pre class="r"><code>germ2snps %&gt;% 
  count(germplasmName) %&gt;% arrange(desc(n))</code></pre>
<pre><code># A tibble: 4,067 × 2
   germplasmName       n
   &lt;chr&gt;           &lt;int&gt;
 1 NR16C1bF185P001     4
 2 NR16F185C1bP001     4
 3 NR16C1bF170P019     3
 4 NR16C1bF170P020     3
 5 NR16C1bF180P002     3
 6 NR16F180C1bP002     3
 7 NR16F182C1bP001     3
 8 NR16C1bF171P002     2
 9 NR16C1bF179P001     2
10 NR16C1bF179P002     2
# ℹ 4,057 more rows</code></pre>
<pre class="r"><code>germ2snps %&gt;% 
  count(FullSampleName) %&gt;% arrange(desc(n))</code></pre>
<pre><code># A tibble: 2,193 × 2
   FullSampleName                     n
   &lt;chr&gt;                          &lt;int&gt;
 1 &lt;NA&gt;                            1747
 2 F116C1bP006:CA7RRANXX:3:499841     3
 3 F123C1bP002:CA7RRANXX:4:499895     3
 4 F123C1bP006:CA7RRANXX:4:499899     3
 5 F12C1bP018:CABV7ANXX:5:503932      3
 6 F135C1bP001:CABJYANXX:5:505238     3
 7 F136C1bP002:CABJYANXX:5:505240     3
 8 F140C1bP001:CABJYANXX:5:505270     3
 9 F152C1bP010:CA7RRANXX:5:499961     3
10 F152C1bP014:CA7RRANXX:5:499965     3
# ℹ 2,183 more rows</code></pre>
<pre class="r"><code>length(unique(dbdata$FullSampleName)) # [1] </code></pre>
<pre><code>[1] 3967</code></pre>
<pre class="r"><code>table(unique(dbdata$FullSampleName) %in% rownames_snps)</code></pre>
<pre><code>
FALSE  TRUE 
 2585  1382 </code></pre>
<pre class="r"><code># FALSE  TRUE 
#  
dbdata %&gt;% 
  select(-GID,-FullSampleName) %&gt;% 
  left_join(germ2snps) %$%
  length(unique(FullSampleName)) # [1] </code></pre>
<pre><code>Joining with `by = join_by(germplasmName)`</code></pre>
<pre><code>Warning in left_join(., germ2snps): Detected an unexpected many-to-many relationship between `x` and `y`.
ℹ Row 19389 of `x` matches multiple rows in `y`.
ℹ Row 1 of `y` matches multiple rows in `x`.
ℹ If a many-to-many relationship is expected, set `relationship =
  &quot;many-to-many&quot;` to silence this warning.</code></pre>
<pre><code>[1] 2193</code></pre>
<pre class="r"><code>dbdata %&gt;% 
  select(-GID,-FullSampleName) %&gt;% 
  left_join(germ2snps) %$%
  table(unique(FullSampleName) %in% rownames_snps) </code></pre>
<pre><code>Joining with `by = join_by(germplasmName)`</code></pre>
<pre><code>Warning in left_join(., germ2snps): Detected an unexpected many-to-many relationship between `x` and `y`.
ℹ Row 19389 of `x` matches multiple rows in `y`.
ℹ Row 1 of `y` matches multiple rows in `x`.
ℹ If a many-to-many relationship is expected, set `relationship =
  &quot;many-to-many&quot;` to silence this warning.</code></pre>
<pre><code>
FALSE  TRUE 
    1  2192 </code></pre>
<pre class="r"><code># FALSE  TRUE 
#    

# Merge updated pheno-to-SNP matches to raw pheno DF
dbdata %&lt;&gt;% 
  select(-GID,-FullSampleName) %&gt;% 
  left_join(germ2snps) %&gt;% 
  # Re-create the GID identifier
  ## Equals the value SNP data name (FullSampleName) 
  ## else germplasmName if no SNP data
  mutate(GID=ifelse(is.na(FullSampleName),germplasmName,FullSampleName))</code></pre>
<pre><code>Joining with `by = join_by(germplasmName)`</code></pre>
<pre><code>Warning in left_join(., germ2snps): Detected an unexpected many-to-many relationship between `x` and `y`.
ℹ Row 19389 of `x` matches multiple rows in `y`.
ℹ Row 1 of `y` matches multiple rows in `x`.
ℹ If a many-to-many relationship is expected, set `relationship =
  &quot;many-to-many&quot;` to silence this warning.</code></pre>
<div id="write-lists-for-matching-genos-to-phenos"
class="section level3">
<h3>Write lists for matching genos-to-phenos</h3>
<pre class="r"><code>dbdata %&gt;% 
  distinct(GID,germplasmName,FullSampleName) %&gt;% 
  write.csv(.,
            file = here::here(&quot;output&quot;,&quot;OnlyChosen_germplasmName_to_FullSampleName_matches_NRCRI_2022Jul09.csv&quot;), 
            row.names = F)</code></pre>
</div>
<div id="output-cleaned-file" class="section level3">
<h3>Output “cleaned” file</h3>
<pre class="r"><code>saveRDS(dbdata,file=here::here(&quot;output&quot;,&quot;NRCRI_CleanedTrialData_2022Jul09.rds&quot;))</code></pre>
</div>
</div>
<div id="detect-experimental-designs" class="section level2">
<h2>Detect experimental designs</h2>
<p>The next step is to check the experimental design of each trial. If
you are absolutely certain of the usage of the design variables in your
dataset, you might not need this step.</p>
<p>Examples of reasons to do the step below:</p>
<p>Some trials appear to be complete blocked designs and the blockNumber
is used instead of replicate, which is what most use. Some complete
block designs have nested, incomplete sub-blocks, others simply copy the
“replicate” variable into the “blockNumber variable” Some trials have
only incomplete blocks but the incomplete block info might be in the
replicate and/or the blockNumber column One reason it might be important
to get this right is that the variance among complete blocks might not
be the same among incomplete blocks. If we treat a mixture of complete
and incomplete blocks as part of the same random-effect
(replicated-within-trial), we assume they have the same variance.</p>
<p>Also error variances might be heterogeneous among different
trial-types (blocking scheme available) and/or plot sizes
(maxNOHAV).</p>
<p>Start with cleaned data from previous step.</p>
<pre class="r"><code>rm(list=ls()); gc()</code></pre>
<pre><code>          used (Mb) gc trigger   (Mb) limit (Mb)  max used   (Mb)
Ncells 1077387 57.6    1803035   96.3         NA   1803035   96.3
Vcells 4362563 33.3  614190782 4686.0      16384 798370683 6091.1</code></pre>
<pre class="r"><code>library(tidyverse); library(magrittr);library(genomicMateSelectR)</code></pre>
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ forcats   1.0.0     ✔ readr     2.1.4
✔ ggplot2   3.4.2     ✔ stringr   1.5.0
✔ lubridate 1.9.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors

Attaching package: &#39;magrittr&#39;


The following object is masked from &#39;package:purrr&#39;:

    set_names


The following object is masked from &#39;package:tidyr&#39;:

    extract</code></pre>
<pre class="r"><code>dbdata&lt;-readRDS(here::here(&quot;output&quot;,&quot;NRCRI_CleanedTrialData_2022Jul09.rds&quot;))

dbdata %&gt;% head %&gt;% rmarkdown::paged_table()</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["studyYear"],"name":[1],"type":["int"],"align":["right"]},{"label":["programName"],"name":[2],"type":["chr"],"align":["left"]},{"label":["locationName"],"name":[3],"type":["chr"],"align":["left"]},{"label":["studyName"],"name":[4],"type":["chr"],"align":["left"]},{"label":["studyDesign"],"name":[5],"type":["chr"],"align":["left"]},{"label":["plotWidth"],"name":[6],"type":["dbl"],"align":["right"]},{"label":["plotLength"],"name":[7],"type":["dbl"],"align":["right"]},{"label":["fieldSize"],"name":[8],"type":["dbl"],"align":["right"]},{"label":["plantingDate"],"name":[9],"type":["chr"],"align":["left"]},{"label":["harvestDate"],"name":[10],"type":["chr"],"align":["left"]},{"label":["germplasmName"],"name":[11],"type":["chr"],"align":["left"]},{"label":["observationUnitDbId"],"name":[12],"type":["int"],"align":["right"]},{"label":["replicate"],"name":[13],"type":["int"],"align":["right"]},{"label":["blockNumber"],"name":[14],"type":["int"],"align":["right"]},{"label":["plotNumber"],"name":[15],"type":["int"],"align":["right"]},{"label":["rowNumber"],"name":[16],"type":["int"],"align":["right"]},{"label":["colNumber"],"name":[17],"type":["int"],"align":["right"]},{"label":["entryType"],"name":[18],"type":["chr"],"align":["left"]},{"label":["TrialType"],"name":[19],"type":["chr"],"align":["left"]},{"label":["observationUnitName"],"name":[20],"type":["chr"],"align":["left"]},{"label":["CGM"],"name":[21],"type":["dbl"],"align":["right"]},{"label":["CGMS1"],"name":[22],"type":["dbl"],"align":["right"]},{"label":["CGMS2"],"name":[23],"type":["dbl"],"align":["right"]},{"label":["DMC"],"name":[24],"type":["dbl"],"align":["right"]},{"label":["DMCsg"],"name":[25],"type":["dbl"],"align":["right"]},{"label":["DRY"],"name":[26],"type":["dbl"],"align":["right"]},{"label":["IVig1-5"],"name":[27],"type":["dbl"],"align":["right"]},{"label":["IVig1-7"],"name":[28],"type":["dbl"],"align":["right"]},{"label":["PLTHT"],"name":[29],"type":["dbl"],"align":["right"]},{"label":["BRNHT1"],"name":[30],"type":["dbl"],"align":["right"]},{"label":["NOHAV"],"name":[31],"type":["dbl"],"align":["right"]},{"label":["StC"],"name":[32],"type":["dbl"],"align":["right"]},{"label":["HI"],"name":[33],"type":["dbl"],"align":["right"]},{"label":["PlotSpacing"],"name":[34],"type":["dbl"],"align":["right"]},{"label":["MaxNOHAV"],"name":[35],"type":["dbl"],"align":["right"]},{"label":["MaxNoHAV"],"name":[36],"type":["dbl"],"align":["right"]},{"label":["logFYLD"],"name":[37],"type":["dbl"],"align":["right"]},{"label":["logDRYLD"],"name":[38],"type":["dbl"],"align":["right"]},{"label":["logDRYsgLD"],"name":[39],"type":["dbl"],"align":["right"]},{"label":["logTOPYLD"],"name":[40],"type":["dbl"],"align":["right"]},{"label":["PropNOHAV"],"name":[41],"type":["dbl"],"align":["right"]},{"label":["MCMDS"],"name":[42],"type":["dbl"],"align":["right"]},{"label":["OrigKeyFile"],"name":[43],"type":["chr"],"align":["left"]},{"label":["OriginOfSample"],"name":[44],"type":["chr"],"align":["left"]},{"label":["FullSampleName"],"name":[45],"type":["chr"],"align":["left"]},{"label":["GID"],"name":[46],"type":["chr"],"align":["left"]}],"data":[{"1":"2006","2":"NRCRI","3":"Umudike","4":"06cetEarlyBulkingUM","5":"RCBD","6":"NA","7":"NA","8":"NA","9":"2006-May-02","10":"NA","11":"COB1111","12":"44773","13":"1","14":"1","15":"105","16":"NA","17":"NA","18":"test","19":"CET","20":"COB1111_2006_Umudike","21":"NA","22":"NA","23":"NA","24":"NA","25":"NA","26":"NA","27":"NA","28":"NA","29":"NA","30":"NA","31":"NA","32":"NA","33":"NA","34":"1","35":"NA","36":"5","37":"NA","38":"NA","39":"NA","40":"NA","41":"NA","42":"NaN","43":"NA","44":"NA","45":"NA","46":"COB1111"},{"1":"2006","2":"NRCRI","3":"Umudike","4":"06cetEarlyBulkingUM","5":"RCBD","6":"NA","7":"NA","8":"NA","9":"2006-May-02","10":"NA","11":"COB1115","12":"44795","13":"1","14":"1","15":"76","16":"NA","17":"NA","18":"test","19":"CET","20":"COB1115_2006_Umudike","21":"NA","22":"NA","23":"NA","24":"NA","25":"NA","26":"NA","27":"NA","28":"NA","29":"NA","30":"NA","31":"NA","32":"NA","33":"NA","34":"1","35":"NA","36":"5","37":"NA","38":"NA","39":"NA","40":"NA","41":"NA","42":"NaN","43":"NA","44":"NA","45":"NA","46":"COB1115"},{"1":"2006","2":"NRCRI","3":"Umudike","4":"06cetEarlyBulkingUM","5":"RCBD","6":"NA","7":"NA","8":"NA","9":"2006-May-02","10":"NA","11":"COB1118","12":"44811","13":"1","14":"1","15":"202","16":"NA","17":"NA","18":"test","19":"CET","20":"COB1118_2006_Umudike","21":"NA","22":"NA","23":"NA","24":"NA","25":"NA","26":"NA","27":"NA","28":"NA","29":"NA","30":"NA","31":"NA","32":"NA","33":"NA","34":"1","35":"NA","36":"5","37":"NA","38":"NA","39":"NA","40":"NA","41":"NA","42":"NaN","43":"NA","44":"NA","45":"NA","46":"COB1118"},{"1":"2006","2":"NRCRI","3":"Umudike","4":"06cetEarlyBulkingUM","5":"RCBD","6":"NA","7":"NA","8":"NA","9":"2006-May-02","10":"NA","11":"COB11","12":"44505","13":"1","14":"1","15":"69","16":"NA","17":"NA","18":"test","19":"CET","20":"COB11_2006_Umudike","21":"NA","22":"NA","23":"NA","24":"NA","25":"NA","26":"NA","27":"NA","28":"NA","29":"NA","30":"NA","31":"NA","32":"NA","33":"NA","34":"1","35":"NA","36":"5","37":"NA","38":"NA","39":"NA","40":"NA","41":"NA","42":"NaN","43":"NA","44":"NA","45":"NA","46":"COB11"},{"1":"2006","2":"NRCRI","3":"Umudike","4":"06cetEarlyBulkingUM","5":"RCBD","6":"NA","7":"NA","8":"NA","9":"2006-May-02","10":"NA","11":"COB1126","12":"44813","13":"1","14":"1","15":"80","16":"NA","17":"NA","18":"test","19":"CET","20":"COB1126_2006_Umudike","21":"NA","22":"NA","23":"NA","24":"NA","25":"NA","26":"NA","27":"NA","28":"NA","29":"NA","30":"NA","31":"NA","32":"NA","33":"NA","34":"1","35":"NA","36":"5","37":"NA","38":"NA","39":"NA","40":"NA","41":"NA","42":"NaN","43":"NA","44":"NA","45":"NA","46":"COB1126"},{"1":"2006","2":"NRCRI","3":"Umudike","4":"06cetEarlyBulkingUM","5":"RCBD","6":"NA","7":"NA","8":"NA","9":"2006-May-02","10":"NA","11":"COB113","12":"44867","13":"1","14":"1","15":"48","16":"NA","17":"NA","18":"test","19":"CET","20":"COB113_2006_Umudike","21":"NA","22":"NA","23":"NA","24":"NA","25":"NA","26":"NA","27":"NA","28":"NA","29":"NA","30":"NA","31":"NA","32":"NA","33":"NA","34":"1","35":"NA","36":"5","37":"NA","38":"NA","39":"NA","40":"NA","41":"NA","42":"NaN","43":"NA","44":"NA","45":"NA","46":"COB113"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
<p>Detect design</p>
<pre class="r"><code>dbdata&lt;-detectExptDesigns(dbdata)

dbdata %&gt;% 
    count(programName,CompleteBlocks,IncompleteBlocks) %&gt;% rmarkdown::paged_table()</code></pre>
<div data-pagedtable="false">
<script data-pagedtable-source type="application/json">
{"columns":[{"label":["programName"],"name":[1],"type":["chr"],"align":["left"]},{"label":["CompleteBlocks"],"name":[2],"type":["fct"],"align":["left"]},{"label":["IncompleteBlocks"],"name":[3],"type":["fct"],"align":["left"]},{"label":["n"],"name":[4],"type":["int"],"align":["right"]}],"data":[{"1":"NRCRI","2":"FALSE","3":"FALSE","4":"1274"},{"1":"NRCRI","2":"FALSE","3":"TRUE","4":"10316"},{"1":"NRCRI","2":"TRUE","3":"FALSE","4":"21077"},{"1":"NRCRI","2":"TRUE","3":"TRUE","4":"24724"}],"options":{"columns":{"min":{},"max":[10]},"rows":{"min":[10],"max":[10]},"pages":{}}}
  </script>
</div>
</div>
<div id="output-file" class="section level2">
<h2>Output file</h2>
<pre class="r"><code>saveRDS(dbdata,file=here::here(&quot;output&quot;,&quot;NRCRI_ExptDesignsDetected_2022Jul09.rds&quot;))</code></pre>
</div>
<div id="home" class="section level2">
<h2><a href="index.html">Home</a></h2>
</div>
<div id="last-step" class="section level2">
<h2>Last step</h2>
<p><a href="Imputation.html">Imputation of DArT Data</a></p>
</div>
<div id="next-step" class="section level1">
<h1>Next step</h1>
<ol start="2" style="list-style-type: decimal">
<li><a href="getBlups.html">Get BLUPs combining all trial data</a>:
Combine data from all trait-trials to get BLUPs for downstream genomic
prediction.</li>
</ol>
<ul>
<li>Fit mixed-model to multi-trial dataset and extract BLUPs,
de-regressed BLUPs and weights. Include two rounds of outlier
removal.</li>
</ul>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.2 (2022-10-31)
Platform: aarch64-apple-darwin20 (64-bit)
Running under: macOS Ventura 13.4.1

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.2-arm64/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] magrittr_2.0.3           lubridate_1.9.2          forcats_1.0.0           
 [4] stringr_1.5.0            readr_2.1.4              ggplot2_3.4.2           
 [7] tidyverse_2.0.0          genomicMateSelectR_0.2.0 purrr_1.0.1             
[10] tidyr_1.3.0              dplyr_1.1.2              tibble_3.2.1            

loaded via a namespace (and not attached):
 [1] tidyselect_1.2.0 xfun_0.39        bslib_0.4.2      colorspace_2.1-0
 [5] vctrs_0.6.3      generics_0.1.3   htmltools_0.5.5  yaml_2.3.7      
 [9] utf8_1.2.3       rlang_1.1.1      jquerylib_0.1.4  later_1.3.1     
[13] pillar_1.9.0     glue_1.6.2       withr_2.5.0      lifecycle_1.0.3 
[17] munsell_0.5.0    gtable_0.3.3     workflowr_1.7.0  evaluate_0.21   
[21] knitr_1.42       tzdb_0.3.0       fastmap_1.1.1    httpuv_1.6.9    
[25] fansi_1.0.4      Rcpp_1.0.11      promises_1.2.0.1 scales_1.2.1    
[29] cachem_1.0.8     jsonlite_1.8.7   fs_1.6.2         hms_1.1.3       
[33] digest_0.6.33    stringi_1.7.12   rprojroot_2.0.3  grid_4.2.2      
[37] here_1.0.1       cli_3.6.1        tools_4.2.2      sass_0.4.5      
[41] whisker_0.4.1    pkgconfig_2.0.3  timechange_0.2.0 rmarkdown_2.21  
[45] rstudioapi_0.14  R6_2.5.1         git2r_0.32.0     compiler_4.2.2  </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
