---
title: "PhenoData"
author: "LucianoRogerio"
date: "2022-07-03"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Download of the phenotypic data on Cassavabase

Follow outlined GenomicPredictionChecklist and previous pipeline to process cassavabase data for ultimate genomic prediction.

Below we will clean and format training data.

Inputs: “Raw” field trial data
Expected outputs: “Cleaned” field trial data

### Cassavabase download
Downloaded all NRCRI field trials, with all the traits using the [download tool](https://cassavabase.org/breeders/download).

Yout need to create a list with all the NRCRI trials on the [wizard tool](https://cassavabase.org/breeders/search)
Cassavabase search wizard:

Selected all NRCRI trials currently available. Make a list. Named it NRCRITrials_2022.

Go to Manage –> [Download here](https://cassavabase.org/breeders/download). Download phenotypes (plot-basis only) and meta-data as CSV using the corresponding boxes / drop-downs.

TRIED TO DOWNLOAD META-DATA, BUT DB IS GIVING “SERVER ERROR”
This happened last year too, must have something to do with NRCRI trials in the list?
Store flatfiles, unaltered in directory data/.
Same error noticed by Marnin.

Read DB data directly from the Cassavabase FTP server.

```{r Read Phenotipic data}
library(genomicMateSelectR)
dbdata<-genomicMateSelectR::readDBdata(phenotypeFile = here::here("data",
                                                                  "2022-07-03T214144phenotype_download.csv"))
```


### Group and select trials to analyze


Make TrialType Variable

```{r Estimate Trial type}
makeTrialTypeVarLuc <- function(indata){
  # So far, this function is not very general
  # Can customize this or add lines to grab TrialTypes for each breeding program
  if(unique(indata$programName)=="NRCRI"){
    outdata<-indata %>%
      mutate(TrialType=ifelse(grepl("TP1",studyName,ignore.case = T),"TP1",NA),
             TrialType=ifelse(grepl("TP2",studyName,ignore.case = T),"TP2",TrialType),
             TrialType=ifelse(grepl("C1a",studyName,ignore.case = T),"C1a",TrialType),
             TrialType=ifelse(grepl("C1b",studyName,ignore.case = T),"C1b",TrialType),
             TrialType=ifelse(grepl("C2a",studyName,ignore.case = T),"C2a",TrialType),
             TrialType=ifelse(grepl("C2b",studyName,ignore.case = T),"C2b",TrialType),
             TrialType=ifelse(grepl("C3a",studyName,ignore.case = T),"C3a",TrialType),
             TrialType=ifelse(grepl("C3b",studyName,ignore.case = T),"C3b",TrialType),
             TrialType=ifelse(grepl("clonal",studyName,ignore.case = T),"CET",TrialType),
             TrialType=ifelse(grepl("cet",studyName,ignore.case = T),"CET",TrialType),
             TrialType=ifelse(grepl("CET",studyName,ignore.case = T),"CET",TrialType),
             TrialType=ifelse(grepl("pyt",studyName,ignore.case = T),"PYT",TrialType),
             TrialType=ifelse(grepl("PYT",studyName,ignore.case = T),"PYT",TrialType),
             TrialType=ifelse(grepl("ayt",studyName,ignore.case = T),"AYT",TrialType),
             TrialType=ifelse(grepl("AYT",studyName,ignore.case = T),"AYT",TrialType),
             TrialType=ifelse(grepl("regional",studyName,ignore.case = T),"UYT",TrialType),
             TrialType=ifelse(grepl("uyt",studyName,ignore.case = T),"UYT",TrialType),
             TrialType=ifelse(grepl("UYT",studyName,ignore.case = T),"UYT",TrialType),
             TrialType=ifelse(grepl("NCRP",studyName) & is.na(TrialType),"NCRP",TrialType),
             TrialType=ifelse(grepl("15nextgen60gs-cbUM|crossnblk|crossingblock",studyName,ignore.case = T) &
                                !grepl("CET",studyName),
                              "CrossingBlock",TrialType),
             TrialType=ifelse(grepl("seedling",studyName,ignore.case = T),NA,TrialType)) }

  return(outdata) }

dbdata<-makeTrialTypeVarLuc(dbdata) 
dbdata %>% 
  count(TrialType) %>% rmarkdown::paged_table()
```

### Remove unclassified trials

```{r Remove unclassified trials}
dbdata %<>% 
    filter(!is.na(TrialType)) 
dbdata %>% 
    group_by(programName) %>% 
    summarize(N=n()) %>% rmarkdown::paged_table()
```

### Make abbreviations for renaming


```{r Change traits names}
traitabbrevs<-tribble(~TraitAbbrev,~TraitName,
        "CMD1S", "cassava.mosaic.disease.severity.1.month.evaluation.CO_334.0000191",
        "CMD3S", "cassava.mosaic.disease.severity.3.month.evaluation.CO_334.0000192",
        "CMD6S", "cassava.mosaic.disease.severity.6.month.evaluation.CO_334.0000194",
        "CMD9S", "cassava.mosaic.disease.severity.9.month.evaluation.CO_334.0000193",
        "CGM", "Cassava.green.mite.severity.CO_334.0000033",
        "CGMS1","cassava.green.mite.severity.first.evaluation.CO_334.0000189",
        "CGMS2","cassava.green.mite.severity.second.evaluation.CO_334.0000190",
        "DMC","dry.matter.content.percentage.CO_334.0000092",
        "DMCsg","dry.matter.content.by.specific.gravity.method.CO_334.0000160",
        "DRY", "dry.yield.CO_334.0000014",
        "IVig1-5", "initial.plant.vigor.assessment.1.5.CO_334.0000220",
        "IVig1-7", "initial.vigor.assessment.1.7.CO_334.0000009",
        "PLTHT","plant.height.measurement.in.cm.CO_334.0000018",
        "BRNHT1","first.apical.branch.height.measurement.in.cm.CO_334.0000106",
        "SHTWT","fresh.shoot.weight.measurement.in.kg.per.plot.CO_334.0000016",
        "RTWT","fresh.storage.root.weight.per.plot.CO_334.0000012",
        "RTNO","root.number.counting.CO_334.0000011",
        "TCHART","total.carotenoid.by.chart.1.8.CO_334.0000161",
        "NOHAV","plant.stands.harvested.counting.CO_334.0000010",
        "StC", "starch.content.percentage.CO_334.0000071")

traitabbrevs %>% rmarkdown::paged_table()
```

Run function renameAndSelectCols() to rename columns and remove everything unecessary

```{r Rename traits}
dbdata<-renameAndSelectCols(traitabbrevs,indata=dbdata,customColsToKeep = c("TrialType","observationUnitName"))
```

### Removing unexpected trait values - Quality Control

```{r Remove typos}
dbdata<-dbdata %>% 
  mutate(CMD1S=ifelse(CMD1S<1 | CMD1S>5,NA,CMD1S),
         CMD3S=ifelse(CMD3S<1 | CMD3S>5,NA,CMD3S),
         CMD6S=ifelse(CMD6S<1 | CMD6S>5,NA,CMD6S),
         CMD9S=ifelse(CMD9S<1 | CMD9S>5,NA,CMD9S),
         CGM=ifelse(CGM<1 | CGM>5,NA,CGM),
         CGMS1=ifelse(CGMS1<1 | CGMS1>5,NA,CGMS1),
         CGMS2=ifelse(CGMS2<1 | CGMS2>5,NA,CGMS2),
         DMC=ifelse(DMC>100 | DMC<=0,NA,DMC),
         DMCsg=ifelse(DMCsg>100 | DMCsg<=0,NA,DMCsg),
         DRY = ifelse(DRY==0 | NOHAV==0 | is.na(NOHAV),NA,DRY),
         `IVig1-5`=ifelse(`IVig1-5`<1 | `IVig1-5`>5,NA,`IVig1-5`),
         `IVig1-7`=ifelse(`IVig1-7`<1 | `IVig1-7`>7,NA,`IVig1-7`),
         RTWT=ifelse(RTWT==0 | NOHAV==0 | is.na(NOHAV),NA,RTWT),
         SHTWT=ifelse(SHTWT==0 | NOHAV==0 | is.na(NOHAV),NA,SHTWT),
         NOHAV=ifelse(NOHAV==0,NA,NOHAV),
         NOHAV=ifelse(NOHAV>42,NA,NOHAV),
         StC=ifelse(StC>100 | StC<=0,NA,StC))

```


## Post-QC traits

### Harvest index

```{r Harvest Index}
dbdata<-dbdata %>% 
    mutate(HI=RTWT/(RTWT+SHTWT))

```

### Unit area traits

I anticipate this will not be necessary as it will be computed before or during data upload.

For calculating fresh root yield:

1. PlotSpacing: Area in m2 per plant. plotWidth and plotLength metadata would hypothetically provide this info, but is missing for vast majority of trials. Therefore, use info from Fola.
2. maxNOHAV: Instead of ExpectedNOHAV. Need to know the max number of plants in the area harvested. For some trials, only the inner (or “net”) plot is harvested, therefore the PlantsPerPlot meta-variable will not suffice. Besides, the PlantsPerPlot information is missing for the vast majority of trials. Instead, use observed max(NOHAV) for each trial. We use this plus the PlotSpacing to calc. the area over which the RTWT was measured. During analysis, variation in the actual number of plants harvested will be accounted for.


```{r Check the number of plant harvest per plot}
dbdata<-dbdata %>% 
    mutate(PlotSpacing=ifelse(programName!="IITA",1,
                              ifelse(studyYear<2013,1,
                              ifelse(TrialType %in% c("CET","GeneticGain","ExpCET"),1,0.8))))
maxNOHAV_byStudy<-dbdata %>% 
  group_by(programName,locationName,studyYear,studyName,studyDesign, TrialType) %>% 
  dplyr::summarise(MaxNOHAV=max(NOHAV, na.rm=T)) %>% 
  ungroup() %>% 
  mutate(MaxNOHAV=ifelse(MaxNOHAV=="-Inf",NA,MaxNOHAV),
         MaxNoHAV=ifelse(TrialType=="CET",
                         yes = 5,
                         no = ifelse(TrialType=="PYT",
                                     yes = 10,
                                     no = ifelse(TrialType=="AYT",
                                                 yes = 20,
                                                 no = ifelse(TrialType=="UYT",
                                                             yes = 36,
                                                             no = MaxNOHAV)))))

write.csv(maxNOHAV_byStudy %>% arrange(studyYear),file=here::here("output","maxNOHAV_byStudy.csv"), row.names = F)
```


This part is copied from Marnin 2021 NRCRI prediction

Previously, I took these values as is. I am unsatisfied with that. The trial number is small enough I’m going to curate manually below. I hope this gives better yield results.

```{r}
maxNOHAV_byStudy %<>%
  mutate(MaxNOHAV=ifelse(studyName=="18C2acrossingblockCETubiaja",8,MaxNOHAV),
         MaxNOHAV=ifelse(studyName=="13TP1CET518kano",5,MaxNOHAV),
         MaxNOHAV=ifelse(studyName=="17C1aAYTGSkano",10,MaxNOHAV),
         MaxNOHAV=ifelse(studyName=="18C1bAYTGSOtobi",10,MaxNOHAV),
         MaxNOHAV=ifelse(studyName=="16C1aCETnonGSOtobi",5,MaxNOHAV),
         MaxNOHAV=ifelse(studyName=="17C1bCETkano",5,MaxNOHAV),
         MaxNOHAV=ifelse(studyName=="16C1aCETnonGSOtobi",5,MaxNOHAV),
         MaxNOHAV=ifelse(studyName=="18C1bAYTGSset2umu",10,MaxNOHAV))
```

Now back to the standart workflow

```{r}
dbdata<-left_join(dbdata,maxNOHAV_byStudy) %>% 
  mutate(RTWT=ifelse(NOHAV>MaxNOHAV,NA,RTWT),
         SHTWT=ifelse(NOHAV>MaxNOHAV,NA,SHTWT),
         HI=ifelse(NOHAV>MaxNOHAV,NA,HI),
         FYLD=RTWT/(MaxNOHAV*PlotSpacing)*10,
         DRYLD=FYLD*(DMC/100),
         DRYsgLD=FYLD*(DMCsg/100),
         logFYLD=log(FYLD),
         logDRYLD=log(DRYLD),
         logDRYsgLD=log(DRYsgLD),
         logTOPYLD=log(SHTWT/(MaxNOHAV*PlotSpacing)*10),
         PropNOHAV=NOHAV/MaxNOHAV) 

# remove non transformed / per-plot (instead of per area) traits
dbdata %<>% select(-RTWT,-SHTWT,-FYLD,-DRYLD,-DRYsgLD)
```

### Season-wide mean disease severity

```{r}
dbdata<-dbdata %>% 
  mutate(MCMDS=rowMeans(.[,colnames(.) %in% c("CMD1S","CMD3S","CMD6S","CMD9S")], na.rm = T)) %>% 
  select(-any_of(c("CMD1S","CMD3S","CMD6S","CMD9S")))

write.table(unique(dbdata$germplasmName), file = here::here("data", "NRCRIClones.txt"),
            quote = F, row.names = F, col.names = F)
```



## [User input] Assign genos to phenos
This step is mostly copy-pasted from previous processing of IITA- and NRCRI-specific data.

Uses 4 flat files, which are available e.g. here. Specifically, `IITA_GBStoPhenoMaster_33018.csv`, `GBSdataMasterList_31818.csv` and `NRCRI_GBStoPhenoMaster_40318.csv` and `chr1_RefPanelAndGSprogeny_ReadyForGP_72719.fam`. I copy them to the `data/` sub-directory for the current analysis.

I downloaded all the Genotyping plates related to the DArT report `DCas-22_7085` (30 genotyping plates).

In addition, DArT-only samples are now expected to also have phenotypes. Therefore, checking for matches in new flatfiles, deposited in the data/ (see code below).

```{r Checking names, eval = F}
library(tidyverse); library(magrittr)
library(here)

DArTNames2018 <- dir(here::here("data", "GenotypingPlates", "C2"))
DArTNames2020 <- dir(here::here("data", "GenotypingPlates", "C3"))
DArTNames2022 <- dir(here::here("data", "GenotypingPlates", "C4"))

GPDArT2018 <- tibble()
for(i in DArTNames2018){
file <- read.table(file = here::here("data", "GenotypingPlates", "C2", i), sep = ",",
                   header = T)

GPDArT2018 <- rbind(GPDArT2018, file)
rm(file)
}

GPDArT2020 <- tibble()
for(i in DArTNames2020){
file <- read.table(file = here::here("data", "GenotypingPlates", "C3", i), sep = ",",
                   header = T)

GPDArT2020 <- rbind(GPDArT2020, file)
rm(file)
}

GPDArT2022 <- tibble()
for(i in DArTNames2022){
file <- read.table(file = here::here("data", "GenotypingPlates", "C4", i), sep = ",",
                   header = T)

GPDArT2022 <- rbind(GPDArT2022, file)
rm(file)
}


GPDArT <- rbind(GPDArT2018, GPDArT2020, GPDArT2022)
write.table(x = GPDArT, file = here::here("data", "NRCRIDArT2022Clones.csv"),
            sep = ",", quote = F, row.names = F)

saveRDS(dbdata, file = here::here("data", "dbdata.rds"))
```

```{r Screening for the Clones genotyped and phenotyped}
dbdata <- readRDS(file = here::here("data", "dbdata.rds"))

gbs2phenoMaster<-dbdata %>% 
  select(germplasmName) %>% 
  distinct %>% 
  left_join(read.csv(here::here("data","NRCRI_GBStoPhenoMaster_40318.csv"), 
                     stringsAsFactors = F)) %>% 
  mutate(FullSampleName=ifelse(grepl("C2a",germplasmName,ignore.case = T) & 
                                 is.na(FullSampleName),germplasmName,FullSampleName)) %>% 
  filter(!is.na(FullSampleName)) %>% 
  select(germplasmName,FullSampleName) %>% 
  bind_rows(dbdata %>% 
              select(germplasmName) %>% 
              distinct %>% 
              left_join(read.csv(here::here("data","IITA_GBStoPhenoMaster_33018.csv"), 
                                 stringsAsFactors = F)) %>% 
              filter(!is.na(FullSampleName)) %>% 
              select(germplasmName,FullSampleName)) %>% 
  bind_rows(dbdata %>% 
              select(germplasmName) %>% 
              distinct %>% 
              left_join(read.csv(here::here("data","GBSdataMasterList_31818.csv"), 
                                 stringsAsFactors = F) %>% 
                          select(DNASample,FullSampleName) %>% 
                          rename(germplasmName=DNASample)) %>% 
              filter(!is.na(FullSampleName)) %>% 
              select(germplasmName,FullSampleName)) %>% 
  bind_rows(dbdata %>% 
              select(germplasmName) %>% 
              distinct %>% 
              mutate(germplasmSynonyms=ifelse(grepl("^UG",germplasmName,ignore.case = T),
                                              gsub("UG","Ug",germplasmName),germplasmName)) %>% 
              left_join(read.csv(here::here("data","GBSdataMasterList_31818.csv"), 
                                 stringsAsFactors = F) %>% 
                          select(DNASample,FullSampleName) %>% 
                          rename(germplasmSynonyms=DNASample)) %>% 
              filter(!is.na(FullSampleName)) %>% 
              select(germplasmName,FullSampleName)) %>%
    bind_rows(dbdata %>% 
              select(germplasmName) %>% 
              distinct %>% 
              left_join(read.csv(here::here("data","NRCRIDArT2022Clones.csv"), 
                                 stringsAsFactors = F) %>% 
                          select(exported_tissue_sample_name, accession_name) %>% 
                          rename(FullSampleName = exported_tissue_sample_name,
                                 germplasmName = accession_name)) %>% 
              filter(!is.na(FullSampleName)) %>% 
              select(germplasmName,FullSampleName)) %>% 
  bind_rows(dbdata %>% 
              select(germplasmName) %>% 
              distinct %>% 
              mutate(germplasmSynonyms=ifelse(grepl("^TZ",germplasmName,
                                                    ignore.case = T),
                                              gsub("TZ","",germplasmName),germplasmName)) %>% 
              left_join(read.csv(here::here("data","GBSdataMasterList_31818.csv"), 
                                 stringsAsFactors = F) %>% 
                          select(DNASample,FullSampleName) %>% 
                          rename(germplasmSynonyms=DNASample)) %>% 
              filter(!is.na(FullSampleName)) %>%
              select(germplasmName,FullSampleName)) %>% 
  distinct %>% 
  left_join(read.csv(here::here("data","GBSdataMasterList_31818.csv"), 
                     stringsAsFactors = F) %>% 
              select(FullSampleName,OrigKeyFile,Institute) %>% 
              rename(OriginOfSample=Institute)) %>% 
  mutate(OrigKeyFile=ifelse(grepl("C2a",germplasmName,ignore.case = T),
                            ifelse(is.na(OrigKeyFile),"LavalGBS",OrigKeyFile),
                            OrigKeyFile),
         OriginOfSample=ifelse(grepl("C2a",germplasmName,ignore.case = T),
                               ifelse(is.na(OriginOfSample),"NRCRI",OriginOfSample),
                               OriginOfSample))

## NEW: check for germName-DArT name matches
germNamesWithoutGBSgenos<-dbdata %>% 
  select(programName,germplasmName) %>% 
  distinct %>% 
  left_join(gbs2phenoMaster) %>% 
  filter(is.na(FullSampleName)) %>% 
  select(-FullSampleName)
## NEW: check for germName-DArT name matches
germNamesWithoutGBSgenos<-dbdata %>% 
  select(programName,germplasmName) %>% 
  distinct %>% 
  left_join(gbs2phenoMaster) %>% 
  filter(is.na(FullSampleName)) %>% 
  select(-FullSampleName)

germNamesWithDArT<-germNamesWithoutGBSgenos %>% 
  inner_join(read.table(here::here("data","chr1_RefPanelAndGSprogeny_ReadyForGP_72719.fam"), 
                        header = F, stringsAsFactors = F)$V2 %>% 
               grep("TMS16|TMS17|TMS18|TMS19|TMS20",.,value = T, ignore.case = T) %>% 
               tibble(dartName=.) %>% 
               separate(dartName,c("germplasmName","dartID"),"_",extra = 'merge',remove = F)) %>% 
  group_by(germplasmName) %>% 
  slice(1) %>% 
  ungroup() %>% 
  rename(FullSampleName=dartName) %>% 
  mutate(OrigKeyFile="DArTseqLD", OriginOfSample="IITA") %>% 
  select(-dartID)
print(paste0(nrow(germNamesWithDArT)," germNames with DArT-only genos"))

# first, filter to just program-DNAorigin matches
germNamesWithGenos<-dbdata %>% 
  select(programName,germplasmName) %>% 
  distinct %>% 
  left_join(gbs2phenoMaster) %>% 
  filter(!is.na(FullSampleName))
print(paste0(nrow(germNamesWithGenos)," germNames with GBS genos"))

# program-germNames with locally sourced GBS samples
germNamesWithGenos_HasLocalSourcedGBS<-germNamesWithGenos %>% 
  filter(programName==OriginOfSample) %>% 
  select(programName,germplasmName) %>% 
  semi_join(germNamesWithGenos,.) %>% 
  group_by(programName,germplasmName) %>% # select one DNA per germplasmName per program
  slice(1) %>% ungroup() 
print(paste0(nrow(germNamesWithGenos_HasLocalSourcedGBS)," germNames with local GBS genos"))

# the rest (program-germNames) with GBS but coming from a different breeding program
germNamesWithGenos_NoLocalSourcedGBS<-germNamesWithGenos %>% 
  filter(programName==OriginOfSample) %>% 
  select(programName,germplasmName) %>% 
  anti_join(germNamesWithGenos,.) %>% 
  # select one DNA per germplasmName per program
  group_by(programName,germplasmName) %>% 
  slice(1) %>% ungroup() 
print(paste0(nrow(germNamesWithGenos_NoLocalSourcedGBS)," germNames without local GBS genos"))

genosForPhenos<-bind_rows(germNamesWithGenos_HasLocalSourcedGBS,
                        germNamesWithGenos_NoLocalSourcedGBS) %>% 
  bind_rows(germNamesWithDArT)

print(paste0(nrow(genosForPhenos)," total germNames with genos either GBS or DArT"))

dbdata %<>% 
    left_join(genosForPhenos) 

# Create a new identifier, GID
## Equals the value SNP data name (FullSampleName) 
## else germplasmName if no SNP data
dbdata %<>% 
  mutate(GID=ifelse(is.na(FullSampleName),germplasmName,FullSampleName))

#  going to check against SNP data
snps<-readRDS(file = here::here("output",
                                "DosageMatrix_DCas22_7085RefPopImputed.rds"))
rownames_snps<-rownames(snps); rm(snps); gc()

dbdata %>% 
  distinct(GID,germplasmName,FullSampleName) %>% 
  semi_join(tibble(GID=rownames_snps)) %>% nrow() #

dbdata %>% 
  distinct(GID,germplasmName,FullSampleName) %>% 
  semi_join(tibble(GID=rownames_snps)) %>% 
  filter(grepl("C1",GID,ignore.case = F)) # 55 C1 clones currently match

dbdata %>% 
  distinct(GID,germplasmName,FullSampleName) %>% 
  semi_join(tibble(GID=rownames_snps)) %>% 
  filter(grepl("C2",GID,ignore.case = F)) # 57 C2 clones either

dbdata %>% 
  distinct(GID,germplasmName,FullSampleName) %>% 
  semi_join(tibble(GID=rownames_snps)) %>% 
  filter(grepl("C3",GID,ignore.case = F)) # no C3 clones either


dbdata %>% 
  distinct(GID,germplasmName,FullSampleName) %>% 
  anti_join(tibble(GID=rownames_snps)) %>% 
  filter(grepl("C1|C2|C3",GID,ignore.case = T)) # definitely there are both C1, C2, and C3 phenotypes

# and there are C1, C2, and C3 genotypes
rownames_snps %>% grep("C1",.,value = T,ignore.case = T) %>% length # [1] 

rownames_snps %>% grep("C2",.,value = T,ignore.case = T) %>% length # [1]  

rownames_snps %>% grep("C3",.,value = T,ignore.case = T) %>% length # [1]  

germ2snps<-dbdata %>% 
  distinct(germplasmName,FullSampleName) %>% 
  semi_join(tibble(FullSampleName=rownames_snps)) %>% 
  bind_rows(dbdata %>% 
              distinct(germplasmName,FullSampleName) %>% 
              anti_join(tibble(FullSampleName=rownames_snps)) %>% 
              filter(grepl("c1a",germplasmName,ignore.case = T)) %>% 
              select(-FullSampleName) %>% 
              left_join(tibble(FullSampleName=rownames_snps) %>% 
                          filter(grepl("c1a",FullSampleName,ignore.case = T)) %>% 
                          separate(FullSampleName,c("dartID","germplasmName"),"\\.\\.\\.",extra = 'merge',remove = F) %>% 
                          select(-dartID))) %>% 
  bind_rows(dbdata %>% 
              distinct(germplasmName,FullSampleName) %>% 
              anti_join(tibble(FullSampleName=rownames_snps)) %>% 
              filter(grepl("C1b",germplasmName,ignore.case = T)) %>% 
              filter(grepl("NR16C1b",germplasmName,ignore.case = T)) %>% 
              select(-FullSampleName) %>%
              left_join(tibble(FullSampleName=rownames_snps) %>% 
                          filter(grepl("c1b",FullSampleName,ignore.case = T)) %>% 
                          separate(FullSampleName,c("germplasmName","GBS_ID"),":",extra = 'merge',remove = F) %>% 
                          select(-GBS_ID) %>% 
                          mutate(germplasmName=gsub("C1b","",germplasmName),
                                 germplasmName=paste0("NR16C1b",germplasmName)))) %>% 
  bind_rows(dbdata %>% 
              distinct(germplasmName,FullSampleName) %>% 
              anti_join(tibble(FullSampleName=rownames_snps)) %>% 
              filter(grepl("C1b",germplasmName,ignore.case = T)) %>% 
              filter(!grepl("NR16C1b",germplasmName,ignore.case = T)) %>% 
              select(-FullSampleName) %>% 
              left_join(tibble(FullSampleName=rownames_snps) %>% 
                          filter(grepl("c1b",FullSampleName,ignore.case = T)) %>% 
                          separate(FullSampleName,c("germplasmName","GBS_ID"),":",extra = 'merge',remove = F) %>% 
                          select(-GBS_ID) %>% 
                          mutate(germplasmName=paste0("NR16",germplasmName)))) %>% 
  bind_rows(dbdata %>% 
              distinct(germplasmName,FullSampleName) %>% 
              anti_join(tibble(FullSampleName=rownames_snps)) %>% 
              filter(grepl("c2",germplasmName,ignore.case = T)) %>% 
              select(-FullSampleName) %>% 
              left_join(tibble(FullSampleName=rownames_snps) %>% 
                          filter(grepl("c2",FullSampleName,ignore.case = T),
                                 grepl("\\.\\.\\.",FullSampleName)) %>% 
                          separate(FullSampleName,c("dartID","germplasmName"),"\\.\\.\\.",extra = 'merge',remove = F) %>% 
                          select(-dartID))) %>% 
  distinct
germ2snps %>% 
  count(germplasmName) %>% arrange(desc(n))

germ2snps %>% 
  count(FullSampleName) %>% arrange(desc(n))

length(unique(dbdata$FullSampleName)) # [1] 

table(unique(dbdata$FullSampleName) %in% rownames_snps)

# FALSE  TRUE 
#  
dbdata %>% 
  select(-GID,-FullSampleName) %>% 
  left_join(germ2snps) %$%
  length(unique(FullSampleName)) # [1] 

dbdata %>% 
  select(-GID,-FullSampleName) %>% 
  left_join(germ2snps) %$%
  table(unique(FullSampleName) %in% rownames_snps) 

# FALSE  TRUE 
#    

# Merge updated pheno-to-SNP matches to raw pheno DF
dbdata %<>% 
  select(-GID,-FullSampleName) %>% 
  left_join(germ2snps) %>% 
  # Re-create the GID identifier
  ## Equals the value SNP data name (FullSampleName) 
  ## else germplasmName if no SNP data
  mutate(GID=ifelse(is.na(FullSampleName),germplasmName,FullSampleName))
```

### Write lists for matching genos-to-phenos

```{r write Clones chosen to GP, eval = F}
dbdata %>% 
  distinct(GID,germplasmName,FullSampleName) %>% 
  write.csv(.,
            file = here::here("output","OnlyChosen_germplasmName_to_FullSampleName_matches_NRCRI_2022Jul09.csv"), 
            row.names = F)
```

### Output “cleaned” file

```{r Saving phenotypic data, eval = F}
saveRDS(dbdata,file=here::here("output","NRCRI_CleanedTrialData_2022Jul09.rds"))
```

## Detect experimental designs

The next step is to check the experimental design of each trial. If you are absolutely certain of the usage of the design variables in your dataset, you might not need this step.

Examples of reasons to do the step below:

Some trials appear to be complete blocked designs and the blockNumber is used instead of replicate, which is what most use.
Some complete block designs have nested, incomplete sub-blocks, others simply copy the “replicate” variable into the “blockNumber variable”
Some trials have only incomplete blocks but the incomplete block info might be in the replicate and/or the blockNumber column
One reason it might be important to get this right is that the variance among complete blocks might not be the same among incomplete blocks. If we treat a mixture of complete and incomplete blocks as part of the same random-effect (replicated-within-trial), we assume they have the same variance.

Also error variances might be heterogeneous among different trial-types (blocking scheme available) and/or plot sizes (maxNOHAV).

Start with cleaned data from previous step.

```{r Read data file}
rm(list=ls()); gc()

library(tidyverse); library(magrittr);library(genomicMateSelectR)

dbdata<-readRDS(here::here("output","NRCRI_CleanedTrialData_2022Jul09.rds"))

dbdata %>% head %>% rmarkdown::paged_table()
```

Detect design

```{r Detect design}
dbdata<-detectExptDesigns(dbdata)

dbdata %>% 
    count(programName,CompleteBlocks,IncompleteBlocks) %>% rmarkdown::paged_table()
```


## Output file

```{r Detect ExptDesign file, eval = F}
saveRDS(dbdata,file=here::here("output","NRCRI_ExptDesignsDetected_2022Jul09.rds"))
```

## [Home](index.html)

## Last step
[Imputation of DArT Data](Imputation.html)

# Next step
2. [Get BLUPs combining all trial data](getBlups.html): Combine data from all trait-trials to get BLUPs for downstream genomic prediction.

- Fit mixed-model to multi-trial dataset and extract BLUPs, de-regressed BLUPs and weights. Include two rounds of outlier removal.